lib/exclusive-lock.js

```diff
@@ -0,0 +1,157 @@
+'use strict'
+
+const Ajv = require('ajv')
+const schemas = require('./schemas/index.js')
+const jsonParseFilter = require('./json-parse-filter.js')
+
+const DEFAULT_LOCK_TTL_MS = 3000
+const DEFAULT_LOCK_REFRESH_MS = 1000
+const REQUIRED_MIN_MS_DIFF = 500
+
+const ajv = new Ajv()
+const validate = ajv.compile(schemas.instantiation_options)
+
+class ExclusiveLock {
+  #lock_name
+
+  constructor(opts) {
+    opts = opts || {}
+
+    if (!opts.log?.error) {
+      const err = new Error('A pino logger instance is required')
+      err.code = 'EINVAL'
+      throw err
+    }
+
+    this.log = opts.log
+
+    if (!validate(opts)) {
+      const err = new Error('Input validation failed')
+      err.code = 'EINVAL'
+      err.errors = validate.errors.map((e) => {
+        const {instancePath, message} = e
+        return {instancePath, message}
+      })
+      this.log.error(err)
+      throw err
+    }
+
+    const {
+      app_name
+    , log
+    , cache_connection
+    , lock_ttl_ms = DEFAULT_LOCK_TTL_MS
+    , lock_refresh_ms = DEFAULT_LOCK_REFRESH_MS
+    , lock_contents = 1 // Can be specified for more context, e.g. POD_NAME
+    , auto_refresh = true
+    } = opts
+
+    if (typeof cache_connection?.set !== 'function') {
+      const err = new TypeError('A valid cache_connection to a cache store is required')
+      err.code = 'EINVAL'
+      throw err
+    }
+    const diff = lock_ttl_ms - lock_refresh_ms
+    if (diff < REQUIRED_MIN_MS_DIFF) {
+      const err = new RangeError(
+        `lock_refresh_ms must be at least ${REQUIRED_MIN_MS_DIFF}ms less than lock_ttl_ms`
+      )
+      err.code = 'EINVAL'
+      err.meta = {
+        lock_ttl_ms
+      , lock_refresh_ms
+      , diff
+      }
+      throw err
+    }
+
+    this.cache_connection = cache_connection
+    this.app_name = app_name.replace(/[\s\W]+/g, '-')
+    this.log = log
+    this.lock_ttl_ms = lock_ttl_ms
+    this.lock_refresh_ms = lock_refresh_ms
+    this.lock_contents = lock_contents
+    this.auto_refresh = auto_refresh
+    this.#lock_name = `lock-manager:${this.app_name}`
+
+    this.got_lock = false
+    this.is_refreshing = false
+    this.refresh_timer = null
+  }
+
+  get lock_name() {
+    return this.#lock_name
+  }
```

Comment:
I don't think the `lock_` prefix is particularly helpful, it can't ever be referenced out of context.
```javascript
lock.lock_name
```

How many times to I have to type lock before you believe its a lock lock?

Reply:
Naming is hard. Good points

Reply:
changed to `key`